stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -742,6 +742,20 @@ extension ${Self}: BinaryFloatingPoint {
     lhs._value = Builtin.fdiv_FPIEEE${bits}(lhs._value, rhs._value)
   }
 
+  @_alwaysEmitIntoClient
+  public static func ** (_ lhs: Self, _ rhs: Self) -> Self {
+% if bits == 16:
+    Self(_stdlib_powf(.init(lhs), .init(rhs)))
+% else:
+    _stdlib_pow${cFuncSuffix}(lhs, rhs)
+% end
+  }
+
+  @_alwaysEmitIntoClient
+  public static func **= (_ lhs: inout Self, _ rhs: Self) {
+    lhs = lhs ** rhs
+  }
+
```

Comment:
The tricky thing about adding these for real (and @stephentyrone has said this in the forums) is that we then need to talk about whether the `(Self, Self) -> Self` overloads should vend IEEE `pow` (integral exponents get special treatment) or `powr` (derived by considering only `exp(rhs * log(lhs))`).

We would also need to consider if `(Self, Int) -> Self` overloads need to be added at the same time (vending IEEE `pown`), and if so whether they can appropriately be spelled with the same operator for two subtly different functions. The reason it'd have to be talked about now is, if we do decide to overload, then for an integer literal exponent adding the overload later would silently change the behavior of existing code.

Reply:
Swift Numerics uses [`__builtin_pow{,f,l}`][shims] for all pairs of [`pow(_:_:)`][APIs] methods:

* [`Float16: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Float16%2BReal.swift#L143-L158)
* [`Float32: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Float%2BReal.swift#L137-L169)
* [`Float64: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Double%2BReal.swift#L150-L196)
* [`Float80: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Float80%2BReal.swift#L132-L144)

Clang also has:

* [`__builtin_powi{,f,l}`][powi] operations, which may be IEEE `pown`?
* `__builtin_powf16` and `__builtin_sqrtf16`, and some support for [`_Float16`][half].

[shims]: <https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/_NumericsShims/include/_NumericsShims.h#L219-L221>
[APIs]:  <https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/ElementaryFunctions.swift#L285-L301>
[powi]:  <https://github.com/apple/llvm-project/blob/16f47cd20b1e573e25a35cce82b5d68138c0b811/clang/include/clang/Basic/Builtins.def#L165-L172>
[half]:  <https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point>

Reply:
Well I think that's settled then...

Reply:
**Correction:** Swift Numerics uses a `guard` statement, to implement `pow(x, y)` as `exp(y * log(x))`:

```swift
extension Double: Real {

  public static func pow(_ x: Double, _ y: Double) -> Double {
    guard x >= 0 else { return .nan }
    return libm_pow(x, y)
  }
}
```

So I think Clang/C99 is IEEE `pow`? Swift Numerics is IEEE `powr` and `pown`?

| Module     |                       Test | Result | IEEE   |
|:-----------|---------------------------:|-------:|:-------|
| Builtin    | `__builtin_pow(-2.0, 4.0)` | `16.0` | `pow`  |
| Darwin     |           `pow(-2.0, 4.0)` | `16.0` | `pow`  |
| RealModule |    `Double.pow(-2.0, 4.0)` | `.nan` | `powr` |
| RealModule |    `Double.pow(-2.0, 4__)` | `16.0` | `pown` |

Reply:
Yeah, conceptually, that would be how they're aligned then (there are some edge conditions where the IEEE-specified result of `pow` and `powr` differ (`pow(1, .infinity)` being one example)).